In [6]:
# importng libraries
import tensorflow as tf
##tensorflow >2.0
from tensorflow.keras.preprocessing.text import one_hot
import numpy as np

In [2]:
### sentences
sent=[  'the glass of milk',
     'the glass of juice',
     'the cup of tea',
    'I am a good boy',
     'I am a good developer',
     'understand the meaning of words',
     'your videos are good',]
sent

['the glass of milk',
 'the glass of juice',
 'the cup of tea',
 'I am a good boy',
 'I am a good developer',
 'understand the meaning of words',
 'your videos are good']

**Initialising the vocubalary**

In [3]:
### Vocabulary size
voc_size=10000

## One Hot Representation

In [4]:
onehot_repr=[one_hot(wrds,voc_size)for wrds in sent]
print(onehot_repr)

[[2257, 8977, 4379, 4229], [2257, 8977, 4379, 6047], [2257, 845, 4379, 2161], [2501, 5843, 6572, 6150, 6528], [2501, 5843, 6572, 6150, 2399], [4150, 2257, 508, 4379, 7905], [5010, 6633, 9753, 6150]]


* 'the glass of milk'-- [9726, 2946, 3433, 1249]
* 'the glass of juice' -- [9726, 2946, 3433, 6200]

**Here we are getting the index of the word in the vocublary**

## Word Embedding Represntation

In [5]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential

**Pad-Sequence**
* it is used because texts in senteb=nxes are not of equal length that's why we use padding to make senteces length equal

In [8]:
sent_length=8
embedded_docs=pad_sequences(onehot_repr,padding="pre",maxlen=sent_length)
embedded_docs

array([[   0,    0,    0,    0, 2257, 8977, 4379, 4229],
       [   0,    0,    0,    0, 2257, 8977, 4379, 6047],
       [   0,    0,    0,    0, 2257,  845, 4379, 2161],
       [   0,    0,    0, 2501, 5843, 6572, 6150, 6528],
       [   0,    0,    0, 2501, 5843, 6572, 6150, 2399],
       [   0,    0,    0, 4150, 2257,  508, 4379, 7905],
       [   0,    0,    0,    0, 5010, 6633, 9753, 6150]])

**Initializing Dimensions**

In [9]:
dim=10

### Adding Embedding layer
* This layer helps us to convert based on dimension featureize representation

In [10]:
model=Sequential()
model.add(Embedding(voc_size,dim,input_length=sent_length))
model.compile('adam','mse')

In [11]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 8, 10)             100000    
                                                                 
Total params: 100,000
Trainable params: 100,000
Non-trainable params: 0
_________________________________________________________________


**Now in oredr to see how our words are converted**

In [12]:
print(model.predict(embedded_docs))

[[[ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743
   -0.03448542 -0.04807053  0.04398764 -0.04839208 -0.03259517]
  [ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743
   -0.03448542 -0.04807053  0.04398764 -0.04839208 -0.03259517]
  [ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743
   -0.03448542 -0.04807053  0.04398764 -0.04839208 -0.03259517]
  [ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743
   -0.03448542 -0.04807053  0.04398764 -0.04839208 -0.03259517]
  [ 0.04591629  0.00305986 -0.00031906 -0.03668549  0.00901998
   -0.0423079  -0.02693666 -0.01885586 -0.04187686  0.00200123]
  [-0.02540212 -0.02978402 -0.04089066  0.03401065  0.04919701
    0.04115206 -0.01142584 -0.04771373 -0.04282392 -0.03050693]
  [ 0.00138439  0.02286117 -0.04833141 -0.01995564  0.04076495
    0.00436453  0.02392674  0.01036782 -0.01952876  0.01613014]
  [ 0.02619321 -0.02836397 -0.00206091  0.00483076 -0.04161619
   -0.02382608  0.04974193  0.01066827  0.041921

In [13]:
## lets see first sentence
embedded_docs[0]

array([   0,    0,    0,    0, 2257, 8977, 4379, 4229])

In [15]:
print(model.predict(embedded_docs)[0])

[[ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743 -0.03448542
  -0.04807053  0.04398764 -0.04839208 -0.03259517]
 [ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743 -0.03448542
  -0.04807053  0.04398764 -0.04839208 -0.03259517]
 [ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743 -0.03448542
  -0.04807053  0.04398764 -0.04839208 -0.03259517]
 [ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743 -0.03448542
  -0.04807053  0.04398764 -0.04839208 -0.03259517]
 [ 0.04591629  0.00305986 -0.00031906 -0.03668549  0.00901998 -0.0423079
  -0.02693666 -0.01885586 -0.04187686  0.00200123]
 [-0.02540212 -0.02978402 -0.04089066  0.03401065  0.04919701  0.04115206
  -0.01142584 -0.04771373 -0.04282392 -0.03050693]
 [ 0.00138439  0.02286117 -0.04833141 -0.01995564  0.04076495  0.00436453
   0.02392674  0.01036782 -0.01952876  0.01613014]
 [ 0.02619321 -0.02836397 -0.00206091  0.00483076 -0.04161619 -0.02382608
   0.04974193  0.01066827  0.04192137 -0.04231681]]


## Note-

* Here each word in sentece is converted into 10 dimensions
* Look at `embedded_docs[0]` first value 4 value is 0, it is converted into [ 0.04188817  0.02073543 -0.04328931 -0.04806497  0.01170743 -0.03448542 -0.04807053  0.04398764 -0.04839208 -0.03259517]
* Means 0 is conveted into 10 values , notice first 4 dimension values are same of model.predict(embedded_docs)[0]
* so embedding basically convert each word into 10 dimensons